packages/cloudflare/src/kv.ts

```diff
@@ -35,70 +35,58 @@ export class KvAdapter<T> implements StorageAdapter<T> {
     return false;
   }
 
-  readAllKeys(): AsyncIterable<string> {
+  async *readAllKeys(): AsyncIterable<string> {
     const kv_namespace = this.kv_namespace;
-    return {
-      async *[Symbol.asyncIterator](): AsyncIterator<string, any, undefined> {
-        let keyList: KVNamespaceListResult<any, string>;
-        const listOptions: KVNamespaceListOptions = {};
-        do {
-          keyList = await kv_namespace.list(listOptions);
-          if (!keyList.list_complete) {
-            listOptions.cursor = keyList.cursor;
-          }
-          for (const key of keyList.keys) {
-            yield key.name;
-          }
-        } while (!keyList.list_complete);
-      },
-    };
+    let keyList: KVNamespaceListResult<any, string>;
+    const listOptions: KVNamespaceListOptions = {};
+    do {
+      keyList = await kv_namespace.list(listOptions);
+      if (!keyList.list_complete) {
+        listOptions.cursor = keyList.cursor;
+      }
+      for (const key of keyList.keys) {
+        yield key.name;
+      }
+    } while (!keyList.list_complete);
```

Comment:
I have a question.

Why do you using while loops to often? Isn't this loop blocking event loop completely while finished?

Reply:
No. `await` does not block anything. Also, `yield` delegates the execution to the caller, so the execution is interrupted there, too.

Let me give you an example. Consider this code:
```ts
function* gen() {
  for (let i = 0; i < 10; i++) {
    console.log("yield", i);
    yield i;
  }
}

let count = 0;
for (const i of gen()) {
  console.log(count++, "receive", i);
}
```
The output of that code is:
```sh
yield 0
0 receive 0
yield 1
1 receive 1
yield 2
2 receive 2
yield 3
3 receive 3
yield 4
4 receive 4
yield 5
5 receive 5
yield 6
6 receive 6
yield 7
7 receive 7
yield 8
8 receive 8
yield 9
9 receive 9
```
Look at how the for loop in `gen` stops executing after each yield. It gives the caller of `gen` the opportunity to process `i` before continuing to iterate. This is why iterator are so powerful.

The article at MDN is actually quite good: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Generator

Reply:
Ok, but what about not generators? Does conversations using generators under the hood, so thats why grammy docs recommended using them?

Reply:
Where do the docs recommend generators?

Reply:
> Where do the docs recommend generators?

I mean docs recommend while loops into converstations.

Reply:
Ah now I see. Well yes, if you do not use generators or async ops then the event loop will be blocked. But for conversations, there are only as many iterations as there are messages in the conversation. Let's assume that the conversation lasts for a few days and the bot and the user exchange 1000 messages. This is already unrealistically high, as most conversations rather take 2-5 messages, but whatever. Then the loop has to go around 1000 times only. This is so fast that you probably cannot even measure it.

Compare this with reading or writing data from the database, or even doing the JSON processing when sending/receiving events from TG. This can easily mean several MB of transferred data per second. Whether or not a small while loop spins a few times is negligible in comparison.

Reply:
Make sense for me now. Thanks for explanation.